### DISCUSSION TITLE
Upload to Debian and Ubuntu repositories?

### DISCUSSION THREAD

**Discussion by fossfreedom** [Ulauncher/Ulauncher#885] [Packaging]
Hi there,

  I am the project lead for Ubuntu Budgie and I'm considering currently whether to add ULauncher as a core component of our distro.  First question - would you be happy with this?  Any reservations that you may have?

The python3 conversion that you have done is a key requirement for us - so it is exciting that you have just started this update.  brilliant news :)

To make ulauncher a key component of UB I will need to create a Debian package that meets current Debian expected standards.  I will then run with this and ask Debian whether they are happy to include your project in their repositories.

There are a number of questions though with the project source and installation structure that need to be addressed though.  Thoughts on the following issues highlighted by Debian lintian?:

1. usr/share/ulauncher/preferences/README.md --> Why is this an installed file?
2. usr/share/ulauncher/preferences/static/fonts/FontAwesome.otf also in fonts-glewlwyd --> why are static fonts being used? Can ulauncher use user defined fonts?
3. font-outside-font-dir usr/share/ulauncher/preferences/static/fonts/FontAwesome.otf --> several font files are included in the source and installed in a odd location (should be /usr/share/fonts) ... similar to 2 - maybe have the ability in the setup package to not install fonts files?
4. binary-without-manpage usr/bin/ulauncher  & binary-without-manpage usr/bin/ulauncher-toggle  --> anything installed in /usr/bin should have the ability to "man ulauncher" and "man ulauncher-toggle"
5. desktop-entry-lacks-keywords-entry usr/share/applications/ulauncher.desktop
6. desktop-command-not-in-package usr/share/applications/ulauncher.desktop env --> would you please consider adding the GDK_BACKEND=x11 environment variable via the python script rather than via the desktop file?
7. debian subfolder - please can you consider moving the debian subfolder and scripts to another branch.  Usually the master branch is just for source files - packaging is handled in other ways such as another branch
8. release versioning. Any chance that in future releases, release versions are tagged "v5.0.0" whereas dev releases are tagged something like "dev xyz" ?  Debian has a concept of watching for new releases. Currently I see v5.0.0r5 which is a dev release will confuse the watch system because it cannot distinguish between a final release and a interim dev drop

**Comment by gornostal** [comment]
Hi @fossfreedom 
I would be very happy to see Ulauncher as a core component of Ubuntu Budgie. 
I have no reservations like that, much less exclusive ones.

The issues you mentioned have one common reason -- I'm not very familiar with packaging for Debian. I find it confusing, TBH. I did packaging configuration mostly by looking at examples, sometimes without a complete understanding of what's going on there. So I could use some help with fixing the issues.

> usr/share/ulauncher/preferences/README.md

No reason for this file to be installed. Should be removed.

> usr/share/ulauncher/preferences/static/fonts/FontAwesome.otf
> font-outside-font-dir 

Ulauncher is not a 100% GTK app. The launcher part is GTK, but the preferences window is an HTML/JS app rendered in a Webkit frame. It requires a custom font from fontawesome.com.
Seems like we have to find a workaround to make Lintian happy about the fonts.

> binary-without-manpage

Fixable. I just have figure out how to add man pages

> desktop-entry-lacks-keywords-entry

Fixable

> desktop-command-not-in-package

I guess I can replace `./bin/ulauncher` with a bash script that execs the python file with `GDK_BACKEND=x11`. So fixable too

> debian subfolder - please can you consider moving the debian subfolder and scripts to another branch

I prefer tracking all changes in the same branch. I can update my release automation to remove that folder before the build.

> release versioning. Any chance that in future releases, release versions are tagged "v5.0.0" 

I think it's a good idea to fix versions, I also don't like the current versioning scheme.

**Comment by gornostal** [comment]
@fossfreedom if Ulauncher is added to the Debian repositories, will I be able to publish new releases whenever I want?

**Comment by fossfreedom** [comment]
Many thanks for the feedback - very encouraging.  cheers.

As to publishing releases - absolutely!  As the maintainer of the package
in Debian and Ubuntu I would be very reliant on releases from your project
and will upload asap when you signal a stable release - or will patch the
package with any significant commits between stable releases where
necessary.

**Comment by fossfreedom** [comment]
For the fonts I think I will need to physically delete the static versions from installing, add a package dependency on fonts-glewlwyd and then symlink the relevant files back into the static folder.  Fiddly - but necessary from a packaging point of view.

I'll try to fix the other issues via the package - and will drop a pull-request for the manpages, desktop-entry-lacks-keywords-entry and desktop-command-not-in-package

Do you have any estimate for when you will adopt a revised version naming convention?  I need this to develop the watch file

**Comment by gornostal** [comment]
> add a package dependency on fonts-glewlwyd and then symlink the relevant files back into the static folder

I really hope there's a simpler solution because with this one we will have to add a lot of extra complexity to the build process.

> I'll try to fix the other issues via the package - and will drop a pull-request for the manpages

That would be great. Thank you :+1: 

> Do you have any estimate for when you will adopt a revised version naming convention? I need this to develop the watch file

I hope I can get this done by the end of the week. If something changes I'll let you know.

**Comment by SwampRabbit** [comment]
@gornostal I assume this got stalled at some point, but I'd like to assist in updating your Debian source packaging files if that is something you'd still like to do.

From what I gathered, I would link any PRs for those commits to this issue?

Debian stages applications in Salsa now (its just GitLab)
https://salsa.debian.org/public
https://wiki.debian.org/Salsa

Another option is the openSUSE Build Service, which will allow you to build for multiple distributions (RHEL, Debian, Ubuntu, openSUSE, Arch, etc) and different architectures.  This is also a bit more self-controlled and less hassle option.
https://build.opensuse.org/

**Comment by gornostal** [comment]
Hi @SwampRabbit 
Thanks for showing your interest in helping with this!

I didn't quite get what commits you were referring to in your question, tbh.
I would be happy to see Ulauncher in the official Debian repo.

The main challenge  with with that is this:
I'm using JS/HTML for Preferences. Re-writing the code to drop JS is not an option -- no one will do it for free :D. 
This means there are build dependencies on nodejs, npm, and probably a few other things so that the final JS file is build from [sources](https://github.com/Ulauncher/Ulauncher/tree/dev/data/preferences/src) and [dozens of NPM packages](https://github.com/Ulauncher/Ulauncher/blob/dev/data/preferences/package.json#L26). 

What can we do about that?
Perhaps we can use a compiled JS file in a debian source package? But AFAIK Debian doesn't allow that.

Or maybe we should extend debian build dependencies with a list that includes NodeJS and other dependencies?
I've considered that too -- not an option either. I'm using a lot of NPM packages that do not exist in debian repos. They only exist in NPM. Even if they existed, they are not the right versions.

Basically to move forward with packaging for official repo we need to find a solution to the question above.
Any ideas?

**Comment by SwampRabbit** [comment]
@gornostal thanks for sharing all that, was very helpful for sure.

I suppose I meant ensuring PR commits are referenced with Issues, like I just did.

Yes it does appear there are a few things that will complicate getting into the official Debian Repos.
I'm still rummaging through ulauncher to fully understand as much as I can, some of it is a bit beyond me in terms of NPM mostly. 

But I'll spend some more time in the next few weeks.  I package quite a few packages, so the more complex and troublesome ones take a bit longer.

Worst case, using openSUSE Build Service is an good option to get ulauncher available for easy install across multiple distros and arches.

**Comment by gornostal** [comment]
I'm more optimistic about the openSUSE Build Service option, tbh.
If it helps with automating releases for different distros then it would be great.

@friday @troycurtisjr what are your thoughts on moving to openSUSE builds service?

**Comment by friday** [comment]
I looked into that not very long ago. It seemed like a good option, but the documentation scared me away.

Does integrating with OBS help the Ubuntu/Debian case compared to the Launchpad packages?

**Comment by SwampRabbit** [comment]
@friday if I understand your question, yes for sure.  
A common misconception is that Debian and Ubuntu packages are the same, in some sense they are, but its like saying that RHEL and Fedora packages are the same.  At compile time different library and package versions get used when building for specific distro releases, building for Ubuntu is more close to building for Debian Testing, not Debian Stable.  OBS builds against each specific distro and version, so if the application builds fine, you know users at least shouldn't have issues installing and less mismatch issues.

Using OBS really isn't hard, its only like maybe a dozen things to set up, after that its really just uploading the new source files and OBS will build for all distros, versions, and architectures you already selected (pending you have the debianization, spec files, and whatever).

Quite a few developers use OBS to distribute their applications for some distros:   Lutris, Palemoon, Polychromatic, onedriver
The MX Linux team uses it sometimes, but we mostly use pbuilder or sbuild and build the old fashioned Debian way.

onedriver recently switched after I mentioned it to them:
https://github.com/jstaf/onedriver/issues/141#issuecomment-898192348
https://software.opensuse.org/download.html?project=home%3Ajstaf&package=onedriver

Some more examples:
https://software.opensuse.org/download.html?project=hardware%3Arazer&package=polychromatic
https://software.opensuse.org/download.html?project=home%3Astrycore&package=lutris

**Comment by friday** [comment]
Thank you for the clarification @SwampRabbit Covering other distros is enough of an advantages with OBS, so I was already hoping to get there at some point, but as I'm sure you could guess there's a lot of other issues with a very high priority and only so much time. At least switching to OBS seems to be something that potentially saves time.

There's already a lot of work done rewriting the setup.py script and other things that affect the package builds, which is going to become part of Ulauncher 6.0.0 at some point. So if you want to keep contributing and make bigger changes, you may want to check first (good idea anyway).
